bigquery/scheduled_queries/calculate-daily-vacancy-tag-time-metrics.sql

```diff
@@ -244,7 +280,20 @@ FROM (
     benchmark_vacancy_metrics
   USING
     (date,
-      tag) )
+      tag)
+  WINDOW
+    dates_before_today AS (
+    PARTITION BY
+      tag
+    ORDER BY
+      date ROWS BETWEEN UNBOUNDED PRECEDING
+      AND CURRENT ROW),
```

Comment:
I've been staring at this bit a while (and googling SQL terms). It sounds like you have a table/query result set of dates (from line 6?) and one of tags (which know what date they were published/expired) and you are 'partitioning' it - which I take to mean (1) grouping the rows by tag (while not reducing the number of rows) and (2) alongside each row inserting the result of a calculation. Then the whole lot of that is ordered by date. How close am I?

Reply:
It's doing a cumulative sum of vacancies published over all time up to a certain date, doing a similar cumulative sum of vacancies expired for the same period and finding the difference, which produces the number of live vacancies on that date. It's using an analytic function: https://cloud.google.com/bigquery/docs/reference/standard-sql/analytic-function-concepts

I'll add a comment that explains this.

Doing it this way avoids having to subquery the whole table each time we want to calculate the number of live vacancies on a particular date, which makes for a much faster query and doesn't exceed the BigQuery maximum subquery limit.